lsコマンドプログラムをオブジェクト指向で作成した

nakamu-kazu222 opened the pull request:
作成した[lsコマンドプログラム](https://github.com/nakamu-kazu222/ruby-practices/blob/myls/04.ls/ls.rb)をオブジェクト指向で作成する
- クラスを分割してクラス同士のやりとりで動作するようにする
- 1 クラスは 1 ファイルに分けて書く

nakamu-kazu222 committed:
ファイル情報の整形と表示を担当するクラスを作成した

nakamu-kazu222 committed:
オプションの処理を担当するクラスを作成した

nakamu-kazu222 committed:
クラスのインスタンスを作成して、lsコマンドを成立させるメインのファイルを作成した

nakamu-kazu222 committed:
ファイル操作を担当するクラスを作成した

nakamu-kazu222 committed:
インデントがずれていたため修正した

yoshitsugu commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,19 @@
+require_relative 'file_permission'
+require_relative 'display'
+
+option = Option.new
+options = option.parse_options
+
+file = FilePermission.new
+array_files = file.list_files(options)
+
+display = Display.new
+array_files = display.align_file_characters(array_files)
+array_of_filenames = display.sort_file_vertical(array_files, options)
+display.display_ls(array_of_filenames, options)
```
現状クラスの使い方が「メソッドの集合」になっていて、あまりオブジェクト指向らしさを感じられないなと思いました。たとえば、 Optionクラスなんかは内部状態の `@options` が必要ないですよね。

実装例として、
```ruby
class LsOption
  def initialize(from)
    @options = parse(from)  
  end

  def show_all_files?
    @options[:a] || false
  end

  # ...
end
```

のようにすると
- `a` オプションが `show_all_files?` のロジックトして隠蔽され、使う側からは `a` オプションを使うと `show_all_fiels?` が trueになる、ということを知らなくてもよい。
  - 将来的にもし `a` オプションが `z` オプションになっても `LsOption` クラスの修正のみで完結する
- 他のオプションに関する要求がでてきても `LsOption` を拡張していけばよさそうなことがわかる
    - たとえば、 `x` と `y` を指定したときだけ何かの機能が有効になる、などの要求がでてきたときも、使う側からはその機能が有効になっているかどうか、だけ意識すればよくそれが `x` や `y` オプションに由来することは `LsOption` だけが知っていればよい、ということになる。

という感じで、責務の分割がすすみ、変更の容易性があがります。
このように概念上存在するモノやコトをオブジェクトとして表現することで、責務の分割をすすめていくのがオブジェクト指向です。再度どういったモノやコトをオブジェクトとして表現すればよいか、考えてみてください。

参考: https://bootcamp.fjord.jp/pages/class-like-class-and-non-class-like-class

nakamu-kazu222 replied:
詳細にご教授いただきありがとうございます

修正の解釈が誤っていたらすみませんが、オプション処理をカプセル化しました

修正したことにより、以下のような実装になったと思います
- Option クラス内でオプションに関するロジックを隠蔽しました
- これにより、ls.rb スクリプトからは具体的なオプション処理の詳細を知らなくてもよくなり、将来的な変更や拡張が容易になりました

yoshitsugu replied:
@nakamu-kazu222 オプションまわりはよくなったと思います :+1: 
この調子で他の箇所もクラス設計を見直してみてください。

nakamu-kazu222 replied:
- FilePermission クラスとDisplay クラスの責務の分割をしました

  - FilePermission クラスは、ファイルに関連する操作を担当するようにしました
    - ファイルリストの取得、ソート、長いフォーマットでのファイル情報の収集など、ファイルに関連する操作を処理します
  
  - Display クラスは表示に関連する操作を担当するようにしました
    - ファイルリストを表示します

yoshitsugu reviewed (commented):
コメントしました。

nakamu-kazu222 committed:
オプション処理をカプセル化した

Option クラス内でオプションに関するロジックを隠蔽している
ls.rb スクリプトからは具体的なオプション処理の詳細を知らなくてもよくなり、将来的な変更や拡張が容易になった

nakamu-kazu222 committed:
クラスの責任を明確にした

FilePermission クラスがファイルに関連する操作を担当するようにした
・ファイルリストの取得、ソート、長いフォーマットでのファイル情報の収集など、ファイルに関連する操作を処理する

Display クラスは表示に関連する操作を担当するようにした
・ファイルリストを表示する

yoshitsugu commented on 07.ls_object/option.rb:
```diff
@@ -0,0 +1,29 @@
+      opts.on('-l') { @options[:l] = true }
+      opts.on('-a') { @options[:a] = true }
+      opts.on('-r') { @options[:r] = true }
+    end.parse(ARGV)
```
細かくいうと、 `ARGV` からのオプションを受け取る場合も考え、`ARGV` はメソッドの引数にしてしまったほうがよいかなと思います。

nakamu-kazu222 replied:
`ARGV` は`parse_options`メソッドの引数に変更しました

yoshitsugu commented on 07.ls_object/file_permission.rb:
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+class FilePermission
```
このクラスが受け持っている責務がファイルのパーミッションだけではないように思います。
命名を考えてみてください。

nakamu-kazu222 replied:
ファイルのパーミッションだけではないので、総括してFileクラスに変更しました

yoshitsugu commented on 07.ls_object/file_permission.rb:
```diff
@@ -0,0 +1,102 @@
+    "#{file_type(name)}#{mode.chars.map { |bit| PERMISSION_SYMBOL[bit] }.join} #{nlink} #{uid} #{gid} #{size} #{mtime} #{name}"
+  end
+
+  def file_status_l_options
```
Optionクラスにoption名は任せているので、 `l_option` という名前がここででてくるのは変ですよね。
命名を考えてみてください。

nakamu-kazu222 replied:
`file_status_l_options`は、optionに関連する名前になっていたので変更しました

yoshitsugu reviewed (commented):
コメントしました。

nakamu-kazu222 committed:
ファイルのパーミッションだけではないので、Fileクラスに変更した

nakamu-kazu222 committed:
ARGV からのオプションを受け取る場合も考え、ARGV はメソッドの引数にした

nakamu-kazu222 committed:
Optionクラスにoption名は任せているので、 l_option という名前を変更した

yoshitsugu commented on 07.ls_object/file.rb:
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+class File
```
`File` クラスというのは既に組込みのものがありますので、衝突しない名前にしましょう。
https://docs.ruby-lang.org/ja/latest/class/File.html
これだとクラスのリオープンになってしまいます。

nakamu-kazu222 replied:
FileOperateクラスに変更しました

yoshitsugu commented on 07.ls_object/file.rb:
```diff
@@ -0,0 +1,102 @@
+
+  def file_type(file)
+    return '-' if File.file?(file)
+    return 'l' if File.symlink?(file)
```
シンボリックリンクも `File.file?` は `true` になりますので、この順番だと `l` があらわれることがなさそうに思いました。

nakamu-kazu222 replied:
シンボリックリンクを区別するために、File.symlink? メソッドを先に評価するようにしました

yoshitsugu commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class Display
```
むずかしいところですが、ファイルを3列の行列に並び変える、という操作も本来のファイル情報としては必要ないので、表示する側の責務ということで、こちらでやるべきな気もしています。どうでしょうか。

nakamu-kazu222 replied:
ファイルを3列の行列に並び変える操作をDisplayクラスに変更しました
FileOperateクラスはデータの処理、Displayクラスは表示という責務に変更しました

yoshitsugu reviewed (commented):
コメントしました。

nakamu-kazu222 committed:
ファイルに対する操作を行うというクラス名に変更した

Fileクラスは、組込みのものがあるので変更した

nakamu-kazu222 committed:
シンボリックリンクを区別するために、File.symlink? メソッドを先に評価するようにした

nakamu-kazu222 committed:
FileOperateクラスはデータの処理、Displayクラスは表示という責務に変更した

yoshitsugu commented on 07.ls_object/file_operate.rb:
```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+class FileOperate
```
クラス名は名詞にしましょう。何らかのモノやコトをあらわすはずなので動詞だと変な感じです。
`FileInformation` とか `FileDetail` とかでいいかなと思います。
そういう意味で、ここでは `file_list` を持つのではなく、個々のファイルに対して1クラスを対応させてもよいかなと思います。

nakamu-kazu222 replied:
FileDetailクラスに変更しました

> そういう意味で、ここでは file_list を持つのではなく、個々のファイルに対して1クラスを対応させてもよいかなと思います。

このコメント内容の解釈として、以下のような修正でよろしいでしょうか？

例えば、FileInfoクラスを用意してFileInfoクラスは、ファイル1つの情報や操作を担当し、
FileDetail クラスは複数の FileInfo インスタンスを管理して全体のファイルを管理する
- FileDetail クラスがディレクトリ内の全てのファイルに対して FileInfo のインスタンスを作成する
- 各 FileInfo インスタンスは、それぞれのファイルに関する情報や操作を持つ
  - 例えば、format_file_status メソッドの中でファイルの情報を収集している部分を FileInfo クラスのメソッドに分ける

yoshitsugu replied:
>  このコメント内容の解釈として、以下のような修正でよろしいでしょうか？

うーん、そうですね。ファイルの配列を担当するクラスと個々のファイルに相当するクラス両方をもつ必要はかならずしもないかなと思っています。
たとえば、

```ruby
class LsCommand
  def main
    show_all, show_reverse, show_long_format = parse_options
    # target_files が FileDetailのArrayになっているイメージ
    target_files = FileDetail.list_by_options(show_all, show_reverse)

    formatter = FileFormatter.new(show_long_format)
    formatter.format(target_files).each do |formatted_line|
        puts formatted_line
    end
  end
end
```

みたいな形でもできるかなと思います。

配列と個々のファイルを担当するクラスを分けるにしても、 `FileDetail` と `FileInfo` だとどちらがどちらなのかわかないので、命名は考えないといけないですね。

nakamu-kazu222 replied:
ご説明のコードを参考にして、変更しました

- LsCommandクラス(旧ls.rb)を追加し、それぞれのクラスで整形された配列を表示するようにしました
  - ls.rbの処理をご説明のコードのようにクラス化して配列を表示するようにしました

- Displayクラスを変更しました
  - DisplayクラスをFileFormatterクラスに名前を変更しました
    - 表示する責務は、LsCommandクラスに移し、Displayクラスはファイルを整形する責務になったので、FileFormatterクラスに変更しました

- ls.rbでは、LsCommandクラスを実行するファイルにしました

yoshitsugu commented on 07.ls_object/file_operate.rb:
```diff
@@ -0,0 +1,90 @@
+    '7' => 'rwx'
+  }.freeze
+
+  def initialize
+    @file_list = Dir.glob('*')
+  end
+
+  def list_files(options)
+    @file_list = Dir.foreach('.').sort if options.show_all_files?
+    @file_list = @file_list.reverse if options.reverse_order?
+    @file_list = format_file_status if options.long_format?
+    @file_list
+  end
```
`list_files` を使わない想定ってありそうですか？
なさそうであれば `initialize` で `list_fiels` でやっていることをやってもいいかもなと思いました。

```ruby
def initialize(show_all_files:, reverse_order:)
  @file_list = #...
end

def format(long_format:)
  align_file_characters(long_format ? format_file_status : @file_list)
end
```

のように対象のファイルをとってくる部分と、それを表示用に加工する部分でわけてもいいかもなと思います。

- `@file_list` にファイル名の配列がはいるパターンとファイル詳細の配列がはいるパターンがあり、 `@file_list` がただのなんでも入る便利変数みたいになっている点
- `ls.rb` で
    ```ruby
    array_files = file_operate.list_files(options)
    array_files = file_operate.align_file_characters(array_files)
    ```
    のようにしてもらいましたが、あまりこの2メソッドを順番通りに呼び出さないといけない必然性がなく、オブジェクトの汎用性としては微妙な点

が気になりました。

nakamu-kazu222 replied:
```
  def initialize(show_all_files:, reverse_order:)
    @file_list = Dir.glob('*')
    @file_list = Dir.foreach('.').sort if show_all_files
    @file_list = @file_list.reverse if reverse_order
  end

  def l_option_formats_file(long_format:)
    align_file_characters(long_format ? format_file_status : @file_list)
  end
```
上記のように修正をしまして、a,rオプション、lオプションの処理を分けました

yoshitsugu commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,29 @@
+class Display
+  COLUMN_COUNT = 3
+
+  def display_ls(array_of_filenames, options)
+    if options.long_format?
+      array_of_filenames.each { |display_filename| puts display_filename }
+    else
+      array_of_filenames.each do |display_filename|
+        display_filename.each do |ls_filename|
+          print ls_filename
+        end
+        print "\n"
+      end
+    end
+  end
```
`option` のうち、 `long_format?` しか使ってないと思うので、そもそも引数としてそれだけしかとらないようにしたほうがよいかなと思います。こういう余計な情報まで渡すことに依存することをスタンプ結合と言ったりします。
https://ja.wikipedia.org/wiki/%E7%B5%90%E5%90%88%E5%BA%A6

nakamu-kazu222 replied:
引数の`options`を`long_format`に修正しました
そのため、long_formatのtrue,falseを判別するようにしました

yoshitsugu commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+  end
+
+  def sort_file_vertical(array_files, options)
```
そもそも `display_ls` の中で呼び出すようにすれば `long_format` の場合は考えなくてよいのではないでしょうか。
オブジェクトの設計の観点からみても、「このメソッドをこの順番で呼び出さないといけない」という暗黙的な知識はできるだけ減らすほうがよいです。

nakamu-kazu222 replied:
ls.rbで順番で呼び出さないようにするため、display.rbのsort_file_vertical メソッド内で display_ls メソッドを呼び出すようにしました

yoshitsugu reviewed (commented):
コメントしました。

nakamu-kazu222 committed:
クラス名を変更した

nakamu-kazu222 committed:
参照するファイル名を変更した

nakamu-kazu222 committed:
a,rオプション、lオプションの処理を分けた

nakamu-kazu222 committed:
optionsは、long_formatのみ判別するので変更した

nakamu-kazu222 committed:
ls.rbで順番で呼び出さないようにした

display.rbのsort_file_vertical メソッド内で display_ls メソッドを呼び出すようにした

nakamu-kazu222 committed:
クラスの責務を変更した。また、DisplayクラスをFileFormatterクラスに変更した

FileFormatterクラスは、ファイル情報を整形するようにした

LsCommandクラスを追加し、それぞれのクラスで整形された配列を表示するようにした

ls.rbでは、LsCommandクラスを実行するファイルにした

nakamu-kazu222 committed:
Rubocopを実行した

yoshitsugu reviewed (approved):
OKです〜
